### PR TITLE
New References - Manage carry over

### PR DESCRIPTION
**Pull request by tomas-stefano**
## Context

When candidates carry over their application they will be moved onto the new references flow. We need to manage the state change of existing references.

**Not submitted applications**
|Old reference status|New reference status|
|---|---|
|Cancelled|Reference deleted|
|Declined|Reference deleted|
|Awaiting response|Not sent|
|Not sent|Not sent|
|Selected|Received (selected isn't a "true" status)|
|Received|Received|

**Unsuccessful (submitted) applications**
|Old reference status|New reference status|
|---|---|
|Cancelled|Reference deleted|
|Declined|Reference deleted|
|Awaiting response|Deleted (not possible to have references that are awaiting response on a submitted application|
|Not sent|Not sent|
|Selected|Received (selected isn't a "true" status)|
|Received|Received|

## Link to Trello card

https://trello.com/c/0IswY7AF/424-references-handling-references-at-carry-over

## Things to check

- [x] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

**Comment by JR-G** [comment]
[References: Handling references at carry over](https://trello.com/c/0IswY7AF/424-references-handling-references-at-carry-over)

File: app/services/duplicate_application.rb [hunk 1]
```diff
@@ -50,16 +50,24 @@ def duplicate
...
+
+    original_references.each do |original_reference|
       new_application_form.application_references.create!(
-        w.attributes.except(*IGNORED_CHILD_ATTRIBUTES).merge!(duplicate: true),
+        original_reference.attributes.except(*IGNORED_CHILD_ATTRIBUTES).merge!(duplicate: true),
       )
 
+      if FeatureFlag.active?(:new_references_flow)
+        awaiting_response_references = new_application_form.application_references.select(&:feedback_requested?)
+        change_references_to_not_requested_yet(awaiting_response_references)
```
Comment (JR-G) [inline app/services/duplicate_application.rb:65]: we also need to mark the references section as incomplete @tomas-stefano

Comment (tomas-stefano) [reply app/services/duplicate_application.rb:65]: Done!

**Review by steventux — CHANGES_REQUESTED** [review]
A couple of comments.

File: app/services/duplicate_application.rb [hunk 2]
```diff
@@ -50,16 +50,25 @@ def duplicate
...
+      .reject(&:feedback_overdue?)
+
+    original_references.each do |original_reference|
       new_application_form.application_references.create!(
-        w.attributes.except(*IGNORED_CHILD_ATTRIBUTES).merge!(duplicate: true),
+        original_reference.attributes.except(*IGNORED_CHILD_ATTRIBUTES).merge!(duplicate: true),
       )
 
+      if FeatureFlag.active?(:new_references_flow)
+        awaiting_response_references = new_application_form.application_references.select(&:feedback_requested?)
```
Comment (carlosmartinez) [inline app/services/duplicate_application.rb:64]: i think rails enum trickery generates class methods such that you can refer to `new_application_form.application_references.feedback_requested`

File: spec/services/duplicate_application_spec.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
+require 'rails_helper'
+
+RSpec.describe DuplicateApplication do
+  subject(:duplicate_application_form) do
+    described_class.new(original_application_form, target_phase: 'apply_2').duplicate
+  end
+
+  let(:original_application_form) do
+    Timecop.travel(-1.day) do
+      @original_application_form ||= create(
```
[the remark this answers is not in the document]
Comment (carlosmartinez) [reply spec/services/duplicate_application_spec.rb:10]: agreed. the time travel, memoization and multiple side-effects within a `let` block feels a bit too complicated to be maintainable.

File: spec/system/candidate_interface/references/candidate_carries_over_unsuccessful_application_to_new_cycle_spec.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    create(
+      :reference,
+      feedback_status: :not_requested_yet,
+      application_form: @application_form,
+    )
+  end
+
+  def when_the_apply2_deadline_passes
+    Timecop.safe_mode = false
```
[the remark this answers is not in the document]
Comment (carlosmartinez) [reply spec/system/candidate_interface/references/candidate_carries_over_unsuccessful_application_to_new_cycle_spec.rb:51]: (that is to say, can we find a way of making this explicit?)

**Comment by carlosmartinez** [comment]
great stuff! just a couple of bits and pieces.

File: app/components/candidate_interface/new_references_section_component.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class CandidateInterface::NewReferencesSectionComponent < ViewComponent::Base
+  include ViewHelper
+
+  attr_reader :application_form_presenter
```
[the remark this answers is not in the document]
Comment (JR-G) [reply app/components/candidate_interface/new_references_section_component.rb:4]: Nah, it was a placeholder to check that the new flow will be present when they carry over.. which reminds me we need to add the `recruitment_cycle_year` check @tomas-stefano

File: spec/system/candidate_interface/references/candidate_carries_over_unsuccessful_application_to_new_cycle_spec.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (JR-G) [reply spec/system/candidate_interface/references/candidate_carries_over_unsuccessful_application_to_new_cycle_spec.rb:51]: @carlosmartinez this isn't one of the specs we added but happy to change as part of this work!

**Comment by JR-G** [comment]
@carlosmartinez & @steventux we've actioned the comments, apart from [this one](https://github.com/DFE-Digital/apply-for-teacher-training/pull/7234#discussion_r931987927)

File: spec/services/duplicate_application_spec.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        volunteering_experiences_count: 1,
+        full_work_history: true,
+        recruitment_cycle_year: RecruitmentCycle.current_year,
+        references_count: 0,
+      )
+      create_list(:reference, 2, feedback_status: :feedback_provided, application_form: @original_application_form)
+      create(:reference, feedback_status: :feedback_refused, application_form: @original_application_form)
+      create(:application_choice, :with_rejection, application_form: @original_application_form)
+    end
+    @original_application_form
```
Comment (carlosmartinez) [inline spec/services/duplicate_application_spec.rb:19]: is this line superfluous?

**Review by carlosmartinez — APPROVED** [review]

File: spec/services/duplicate_application_spec.rb (hunk 6 above)
Comment (JR-G) [reply spec/services/duplicate_application_spec.rb]: 🪓

**Pull request merged**